source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc

```diff
@@ -0,0 +1,202 @@
+      return Http::FilterTrailersStatus::Continue;
+    }
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+Http::FilterHeadersStatus BandwidthLimiter::encodeHeaders(Http::ResponseHeaderMap&, bool) {
+  auto* config = getConfig();
+
+  if (config->enabled() && (config->enableMode() & BandwidthLimit::ENCODE)) {
+    config->stats().encode_enabled_.inc();
+
+    encode_limiter_ = std::make_unique<StreamRateLimiter>(
+        config->limit(), encoder_callbacks_->encoderBufferLimit(),
+        [this] { encoder_callbacks_->onEncoderFilterAboveWriteBufferHighWatermark(); },
+        [this] { encoder_callbacks_->onEncoderFilterBelowWriteBufferLowWatermark(); },
+        [this](Buffer::Instance& data, bool end_stream) {
+          if (end_stream) {
+            updateStatsOnEncodeFinish();
+          }
+          encoder_callbacks_->injectEncodedDataToFilterChain(data, end_stream);
+        },
+        [this] {
+          updateStatsOnEncodeFinish();
+          encoder_callbacks_->continueEncoding();
+        },
+        [config](uint64_t len) { config->stats().encode_allowed_size_.set(len); },
+        const_cast<FilterConfig*>(config)->timeSource(), encoder_callbacks_->dispatcher(),
+        encoder_callbacks_->scope(), config->tokenBucket(), config->fillInterval());
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus BandwidthLimiter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (encode_limiter_ != nullptr) {
+    const auto* config = getConfig();
+
+    if (!encode_latency_) {
+      encode_latency_ = std::make_unique<Stats::HistogramCompletableTimespanImpl>(
+          config->stats().encode_transfer_duration_,
+          const_cast<FilterConfig*>(config)->timeSource());
+      config->stats().encode_pending_.inc();
+    }
+    config->stats().encode_incoming_size_.set(data.length());
+
+    encode_limiter_->writeData(data, end_stream);
+    return Http::FilterDataStatus::StopIterationNoBuffer;
+  }
+  ENVOY_LOG(debug, "BandwidthLimiter <encode data>: encode_limiter not set");
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus BandwidthLimiter::encodeTrailers(Http::ResponseTrailerMap&) {
+  if (encode_limiter_ != nullptr) {
+    if (encode_limiter_->onTrailers()) {
+      return Http::FilterTrailersStatus::StopIteration;
+    } else {
+      updateStatsOnEncodeFinish();
+      return Http::FilterTrailersStatus::Continue;
+    }
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void BandwidthLimiter::updateStatsOnDecodeFinish() {
+  if (decode_latency_) {
+    decode_latency_->complete();
+    decode_latency_.reset();
+    getConfig()->stats().decode_pending_.dec();
+  }
+}
+
+void BandwidthLimiter::updateStatsOnEncodeFinish() {
+  if (encode_latency_) {
+    encode_latency_->complete();
+    encode_latency_.reset();
+    getConfig()->stats().encode_pending_.dec();
+  }
+}
+
+const FilterConfig* BandwidthLimiter::getConfig() const {
```

Comment:
nit: if this cannot return a null, (from code above they assume so, use reference)

Reply:
good idea